File: C++/Palindrome-Reorder.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;


int main() {
	string input;
	cin >> input;

	int c[26] = {}, c1;
	for(char d : input)
		++c[d - 'A'];
	for(int i = 0; i < 26; i++)
		c1 += c[i] & 1;
	if (c1 > 1) {
		cout << "NO SOLUTION";
		return 0;
	}

	// Solution possible
	string t;
	for(int i = 0; i < 26; i++) {
		if (c[i] % 2 == 0) {
			// Even counts
			for (int j = 0; j < c[i] / 2; j++) {
				t += (char)('A' + i);
	}}}

	cout << t;

	for (int i = 0; i < 26; i++) {
		if (c[i] & 1) {
			// Odd count
			for (int j = 0; j < c[i]; j++) {
				cout << (char)('A' + i);
			}
		}
	}

	reverse(t.begin(), t.end());
	cout << t;

	return 0;
}

File: Python/Palindrome-Reorder.py
def reorder(sequence):
    alpha = dict()
    odd_count = 0
    for char in sequence:
        if char in alpha:
            alpha[char] += 1
        else:
            alpha[char] = 1
    for key, values in alpha.items():
        if values % 2 != 0:  # Odd number of characters
            odd_count += 1
    if odd_count > 1:
        print("NO SOLUTION")
        return
    output = None
    for key, values in alpha.items():
        if values % 2 == 0:  # Even count
            for i in range(values // 2):
                if output is None:
                    output = key
                else:
                    output += key
    if output is not None:
        print(output, end='')
    for key, values in alpha.items():
        if values % 2 != 0:
            for i in range(values):
                print(key, end='')
    # Reverse Output
    if output is not None:
        print(output[::-1])

if __name__ == '__main__':
    sequence = input()

    reorder(sequence)

